tensorflow_asr/datasets/asr_dataset.py

```diff
@@ -366,3 +366,344 @@ def create(self, batch_size: int):
+                        if base_index + j < 0 or base_index + j >= num_frames:
+                            continue
+                        mask[i, base_index + j] = 1
+            return mask
+
+        @tf.function(autograph=True)
+        def _calculate_mask_tf(num_frames, frames_per_chunk, history_window_size):
+            chunk_ids = tf.range(num_frames) // frames_per_chunk
+            num_chunks = tf.cast(tf.math.ceil(num_frames / frames_per_chunk), dtype=tf.int32)
+
+            # Create first `frames_per_chunk` rows
+            current = tf.ones((frames_per_chunk), dtype=tf.int32)
+            trailing = tf.ones(((num_chunks - 1) * frames_per_chunk), dtype=tf.int32)
+            tmp_row = tf.concat((current, trailing), axis=0)
+            row = tf.slice(tmp_row, [0], [num_frames])
+            mask = tf.expand_dims(row, axis=0)
+
+            for i in range(1, frames_per_chunk):
+                mask = tf.concat((mask, [row]), axis=0)
+
+            # Create the following rows
+            for i in range(frames_per_chunk, num_frames):
+                tf.autograph.experimental.set_loop_options(
+                    shape_invariants=[(mask, tf.TensorShape([None, None]))]
+                )
+                curr_chunk_id = chunk_ids[i]
+                hist_i = tf.math.maximum(i - history_window_size, 0)
+                hist_chunk_id = chunk_ids[hist_i]
+
+                # Build the left-most part
+                leading_chunk_id = hist_chunk_id - 1
+                num_leading_chunks = tf.math.maximum(leading_chunk_id + 1, 0)
+                leftmost_row = tf.zeros((num_leading_chunks * frames_per_chunk), dtype=tf.int32)
+
+                # Build the current visible chunks
+                num_hist_chunks = curr_chunk_id - hist_chunk_id
+                num_visible_chunks = num_hist_chunks + 1
+                curr_chunk_row = tf.ones((num_visible_chunks * frames_per_chunk), dtype=tf.int32)
+
+                # Build the trailing 0s
+                num_trailing_chunks = tf.math.maximum((num_chunks - curr_chunk_id) - 1, 0)
+                trailing_chunk_row = tf.zeros((num_trailing_chunks * frames_per_chunk), dtype=tf.int32)
+
+                # Merge chunks, clip to output size
+                tmp_row = tf.concat([leftmost_row, curr_chunk_row, trailing_chunk_row], axis=0)
+                row = tf.slice(tmp_row, [0], [num_frames])
+
+                mask = tf.concat((mask, [row]), axis=0)
+            return mask
+
+        if self.use_tf:
+            mask = _calculate_mask_tf(num_frames, frames_per_chunk, self.history_window_size)
+        else:
+            mask = tf.numpy_function(
+                _calculate_mask, inp=[num_frames, frames_per_chunk, self.history_window_size], Tout=tf.int32
+            )
+            mask.set_shape((None, None))
+
+        return mask
+
+    def calculate_mask(self, num_frames):
+        num_frames = tf.cast(tf.math.ceil(num_frames / self.time_reduction_factor), tf.int32)
```

Comment:
@andreselizondo-adestech We should use [get_reduce_length](https://github.com/TensorSpeech/TensorFlowASR/blob/fdd878292e7ce478d7644c4fed1f996792ebee73/tensorflow_asr/utils/math_util.py#L43)

Reply:
Done